services/streammanager/kafka/kafkamanager.go

```diff
@@ -398,14 +431,30 @@ func prepareMessage(topic, key string, message []byte, timestamp time.Time) clie
 	}
 }
 
-func prepareBatchOfMessages(topic string, batch []map[string]interface{}, timestamp time.Time) (
+// This function is used to serialize the binary data according to the avroSchema.
+// It iterates over the schemas provided by the customer and tries to serialize the data.
+// If it's able to serialize the data then it returns the converted data otherwise it returns an error.
+// We are using the LinkedIn goavro library for data serialization. Ref: https://github.com/linkedin/goavro
+func serialize(value []byte, codec goavro.Codec) ([]byte, error) {
+	native, _, err := codec.NativeFromTextual(value)
+	if err != nil {
+		return nil, fmt.Errorf("unable convert the event to native from textual, with error: %s", err)
+	}
+	binary, err := codec.BinaryFromNative(nil, native)
+	if err != nil {
+		return nil, fmt.Errorf("unable convert the event to binary from native, with error: %s", err)
+	}
+	return binary, nil
+}
+
+func prepareBatchOfMessages(topic string, batch []map[string]interface{}, timestamp time.Time, p producer) (
 	[]client.Message, error,
 ) {
 	start := now()
 	defer func() { kafkaStats.prepareBatchTime.SendTiming(since(start)) }()
 
 	var messages []client.Message
-	for _, data := range batch {
+	for i, data := range batch {
```

Comment:
Given the existing logic of this function, please check with product if you should return in this `for` loop when there are codec errors instead of doing `continue`. As you can see, if there is a missing `message` or a missing `userId`, we log and carry on so that we can process the rest of the batch. On the other hand, when you come across a codec error you just return and ignore the rest of the messages in the batch.

Reply:
Anyhow we are discarding the entire batch for Marshall error in line no.471. @arnab-p WDYT?

Reply:
Anyhow, we are discarding the entire batch for any marshal error. @arnab-p WDYT?